File: src/shared/__tests__/sitePatterns.test.ts
import { describe, expect, it } from 'vitest'
import {
  compareSitePatterns,
  matchSitePatterns,
  normalizeSitePattern,
  parseSitePattern,
  siteAddressOf,
  sitePatternCoversHost,
  sitePatternForHost,
  sitePatternMatches,
  urlInSitePatterns,
  type SitePattern
} from '../sitePatterns'

describe('parseSitePattern', () => {
  it("reads Chrome's grammar and writes the canonical text back", () => {
    expect(parseSitePattern('example.com')).toEqual({
      text: 'example.com',
      scheme: null,
      host: 'example.com',
      subdomains: false,
      port: null
    })
    expect(parseSitePattern('[*.]example.com')).toMatchObject({
      text: '[*.]example.com',
      subdomains: true
    })
    expect(parseSitePattern('https://[*.]example.com')).toMatchObject({
      text: 'https://[*.]example.com',
      scheme: 'https',
      subdomains: true
    })
    expect(parseSitePattern('http://example.com:8080')).toMatchObject({
      text: 'http://example.com:8080',
      scheme: 'http',
      port: 8080
    })
    expect(parseSitePattern('[*.]example.com:8080')).toMatchObject({
      text: '[*.]example.com:8080',
      subdomains: true,
      port: 8080
    })
    // Case, whitespace, a trailing dot and the any-scheme / any-port wildcards are forgiven.
    expect(normalizeSitePattern('  HTTPS://Example.COM.  ')).toBe('https://example.com')
    expect(normalizeSitePattern('*://example.com:*')).toBe('example.com')
    expect(normalizeSitePattern('[*.]Example.com:*')).toBe('[*.]example.com')
    // IDN travels as punycode, as URLs carry it.
    expect(normalizeSitePattern('xn--bcher-kva.example')).toBe('xn--bcher-kva.example')
  })

  it('takes IP literals as one machine each', () => {
    expect(normalizeSitePattern('192.168.0.1')).toBe('192.168.0.1')
    expect(normalizeSitePattern('192.168.0.1:8443')).toBe('192.168.0.1:8443')
    expect(normalizeSitePattern('[::1]')).toBe('[::1]')
    expect(normalizeSitePattern('::1')).toBe('[::1]')
    expect(normalizeSitePattern('https://[2001:db8::1]:8443')).toBe('https://[2001:db8::1]:8443')
    // Nothing is under an address: the wildcard is refused.
    expect(parseSitePattern('[*.]192.168.0.1')).toBeNull()
    expect(parseSitePattern('[*.][::1]')).toBeNull()
    // Not an address, not a name either.
    expect(parseSitePattern('999.1.1.1')).toBeNull()
  })

  it('refuses what is not a pattern', () => {
    for (const bad of [
      '',
      '   ',
      'https://',
      'ftp://example.com',
      'example.com/path',
      'https://example.com/',
      'user@example.com',
      'exam ple.com',
      '*.example.com',
      'ex*mple.com',
      '[*.]',
      '-bad.example',
      'bad-.example',
      'example.com:0',
      'example.com:65536',
      'example.com:12a',
      `${'a'.repeat(64)}.example`,
      `${'a.'.repeat(130)}com`
    ])
      expect(parseSitePattern(bad), bad).toBeNull()
  })
})

describe('sitePatternForHost', () => {
  it('gives a bare host the subdomain wildcard and leaves an IP literal exact', () => {
    expect(sitePatternForHost('Example.com')).toBe('[*.]example.com')
    expect(sitePatternForHost('www.example.com')).toBe('[*.]www.example.com')
    expect(sitePatternForHost('10.0.0.2')).toBe('10.0.0.2')
    expect(sitePatternForHost('[::1]')).toBe('[::1]')
    // A pattern with a scheme or port is not a host.
    expect(sitePatternForHost('https://example.com')).toBeNull()
    expect(sitePatternForHost('example.com:8080')).toBeNull()
    expect(sitePatternForHost('not a host')).toBeNull()
  })
})

describe('siteAddressOf', () => {
  it("takes the scheme, host and port of a URL, the scheme's default port without one", () => {
    expect(siteAddressOf('https://Example.com/a?b#c')).toEqual({
      scheme: 'https',
      host: 'example.com',
      port: 443
    })
    expect(siteAddressOf('http://example.com:8080/')).toEqual({
      scheme: 'http',
      host: 'example.com',
      port: 8080
    })
    // WebSocket URLs are looked up as the http scheme they ride on.
    expect(siteAddressOf('wss://live.example/socket')).toEqual({
      scheme: 'https',
      host: 'live.example',
      port: 443
    })
    expect(siteAddressOf('ws://live.example/socket')?.scheme).toBe('http')
    expect(siteAddressOf('https://[::1]:8443/')).toEqual({
      scheme: 'https',
      host: '[::1]',
      port: 8443
    })
    expect(siteAddressOf('about:blank')).toBeNull()
    expect(siteAddressOf('not a url')).toBeNull()
    expect(siteAddressOf({ scheme: 'https', host: 'a.example', port: null })).toEqual({
      scheme: 'https',
      host: 'a.example',
      port: null
    })
  })
})

describe('sitePatternMatches', () => {
  const p = (text: string): SitePattern => parseSitePattern(text)!

  it('covers the host alone without [*.] and its subdomains with it', () => {
    expect(sitePatternMatches(p('example.com'), 'https://example.com/')).toBe(true)
    expect(sitePatternMatches(p('example.com'), 'https://www.example.com/')).toBe(false)
    expect(sitePatternMatches(p('[*.]example.com'), 'https://example.com/')).toBe(true)
    expect(sitePatternMatches(p('[*.]example.com'), 'https://a.b.example.com/')).toBe(true)
    expect(sitePatternMatches(p('[*.]example.com'), 'https://notexample.com/')).toBe(false)
    expect(sitePatternMatches(p('[*.]example.com'), 'https://example.com.evil/')).toBe(false)
    expect(sitePatternCoversHost(p('[*.]example.com'), 'Sub.Example.com.')).toBe(true)
    expect(sitePatternCoversHost(p('example.com'), 'sub.example.com')).toBe(false)
  })

  it('narrows to a scheme or a port when the pattern names one', () => {
    expect(sitePatternMatches(p('https://example.com'), 'https://example.com/')).toBe(true)
    expect(sitePatternMatches(p('https://example.com'), 'http://example.com/')).toBe(false)
    expect(sitePatternMatches(p('https://example.com'), 'wss://example.com/')).toBe(true)
    expect(sitePatternMatches(p('example.com:8443'), 'https://example.com:8443/')).toBe(true)
    expect(sitePatternMatches(p('example.com:8443'), 'https://example.com/')).toBe(false)
    // A pattern without a port matches the default port and any other.
    expect(sitePatternMatches(p('example.com'), 'https://example.com:8443/')).toBe(true)
    expect(sitePatternMatches(p('example.com:443'), 'https://example.com/')).toBe(true)
    expect(sitePatternMatches(p('[::1]'), 'http://[::1]:3000/')).toBe(true)
    expect(sitePatternMatches(p('example.com'), 'about:blank')).toBe(false)
  })
})

describe('compareSitePatterns and matchSitePatterns', () => {
  it('orders from the most specific: exact host, longer host, named scheme, named port', () => {
    const order = [
      'https://www.example.com:8443',
      'https://www.example.com',
      'www.example.com:8443',
      'www.example.com',
      'example.com',
      'https://[*.]www.example.com',
      '[*.]www.example.com:8443',
      '[*.]www.example.com',
      '[*.]example.com'
    ]
    const shuffled = [...order].reverse()
    const sorted = shuffled
      .map((t) => parseSitePattern(t)!)
      .sort(compareSitePatterns)
      .map((p) => p.text)
    expect(sorted).toEqual(order)
    // Ties are total (by text) so a list has one order.
    const a = parseSitePattern('a.example')!
    const b = parseSitePattern('b.example')!
    expect(compareSitePatterns(a, b)).toBeLessThan(0)
    expect(compareSitePatterns(b, a)).toBeGreaterThan(0)
    expect(compareSitePatterns(a, a)).toBe(0)
  })

  it('picks the most specific covering pattern of a list and skips what does not parse', () => {
    const list = ['[*.]example.com', 'www.example.com', 'garbage/', 'https://[*.]example.com']
    expect(matchSitePatterns(list, 'https://www.example.com/')?.text).toBe('www.example.com')
    expect(matchSitePatterns(list, 'https://cdn.example.com/')?.text).toBe(
      'https://[*.]example.com'
    )
    expect(matchSitePatterns(list, 'http://cdn.example.com/')?.text).toBe('[*.]example.com')
    expect(matchSitePatterns(list, 'https://other.example/')).toBeNull()
    expect(matchSitePatterns(list, 'about:blank')).toBeNull()
    expect(urlInSitePatterns(list, 'https://example.com/')).toBe(true)
    expect(urlInSitePatterns([], 'https://example.com/')).toBe(false)
  })
})

File: android/app/src/test/kotlin/app/zen/chromium/privacy/SitePatternsTest.kt
package app.zen.chromium.privacy

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Chrome's content-settings pattern grammar; the twin of `src/shared/__tests__/sitePatterns.test.ts`. */
class SitePatternsTest {
    private fun p(text: String): SitePattern = SitePattern.parse(text) ?: error("did not parse: $text")

    @Test
    fun `reads the grammar and writes the canonical text back`() {
        val exact = p("example.com")
        assertEquals("example.com", exact.text)
        assertNull(exact.scheme)
        assertEquals("example.com", exact.host)
        assertFalse(exact.subdomains)
        assertNull(exact.port)
        assertTrue(p("[*.]example.com").subdomains)
        assertEquals("https", p("https://[*.]example.com").scheme)
        assertEquals(8080, p("http://example.com:8080").port)
        assertEquals("[*.]example.com:8080", p("[*.]example.com:8080").text)
        // Case, whitespace, a trailing dot and the any-scheme / any-port wildcards are forgiven.
        assertEquals("https://example.com", SitePattern.normalize("  HTTPS://Example.COM.  "))
        assertEquals("example.com", SitePattern.normalize("*://example.com:*"))
        assertEquals("[*.]example.com", SitePattern.normalize("[*.]Example.com:*"))
        assertEquals("xn--bcher-kva.example", SitePattern.normalize("xn--bcher-kva.example"))
    }

    @Test
    fun `takes IP literals as one machine each`() {
        assertEquals("192.168.0.1", SitePattern.normalize("192.168.0.1"))
        assertEquals("192.168.0.1:8443", SitePattern.normalize("192.168.0.1:8443"))
        assertEquals("[::1]", SitePattern.normalize("[::1]"))
        assertEquals("[::1]", SitePattern.normalize("::1"))
        assertEquals("https://[2001:db8::1]:8443", SitePattern.normalize("https://[2001:db8::1]:8443"))
        assertNull(SitePattern.parse("[*.]192.168.0.1"))
        assertNull(SitePattern.parse("[*.][::1]"))
        assertNull(SitePattern.parse("999.1.1.1"))
    }

    @Test
    fun `refuses what is not a pattern`() {
        for (bad in listOf(
            "", "   ", "https://", "ftp://example.com", "example.com/path", "https://example.com/",
            "user@example.com", "exam ple.com", "*.example.com", "ex*mple.com", "[*.]", "-bad.example",
            "bad-.example", "example.com:0", "example.com:65536", "example.com:12a",
            "a".repeat(64) + ".example", "a.".repeat(130) + "com"
        )) assertNull(bad, SitePattern.parse(bad))
    }

    @Test
    fun `gives a bare host the subdomain wildcard and leaves an IP literal exact`() {
        assertEquals("[*.]example.com", SitePattern.forHost("Example.com"))
        assertEquals("[*.]www.example.com", SitePattern.forHost("www.example.com"))
        assertEquals("10.0.0.2", SitePattern.forHost("10.0.0.2"))
        assertEquals("[::1]", SitePattern.forHost("[::1]"))
        assertNull(SitePattern.forHost("https://example.com"))
        assertNull(SitePattern.forHost("example.com:8080"))
        assertNull(SitePattern.forHost("not a host"))
    }

    @Test
    fun `reads a URL's scheme, host and port, the scheme's default without one`() {
        val a = SiteAddress.of("https://Example.com/a?b#c")!!
        assertEquals(listOf("https", "example.com", 443), listOf(a.scheme, a.host, a.port))
        assertEquals(8080, SiteAddress.of("http://example.com:8080/")!!.port)
        // WebSocket URLs are looked up as the http scheme they ride on.
        val ws = SiteAddress.of("wss://live.example/socket")!!
        assertEquals(listOf("https", "live.example", 443), listOf(ws.scheme, ws.host, ws.port))
        assertEquals("http", SiteAddress.of("ws://live.example/socket")!!.scheme)
        val v6 = SiteAddress.of("https://[::1]:8443/")!!
        assertEquals(listOf("https", "[::1]", 8443), listOf(v6.scheme, v6.host, v6.port))
        assertEquals("example.com", SiteAddress.of("https://user:pw@example.com/")!!.host)
        assertNull(SiteAddress.of("about:blank"))
        assertNull(SiteAddress.of("not a url"))
    }

    @Test
    fun `covers the host alone without the wildcard and its subdomains with it`() {
        assertTrue(p("example.com").matches("https://example.com/"))
        assertFalse(p("example.com").matches("https://www.example.com/"))
        assertTrue(p("[*.]example.com").matches("https://example.com/"))
        assertTrue(p("[*.]example.com").matches("https://a.b.example.com/"))
        assertFalse(p("[*.]example.com").matches("https://notexample.com/"))
        assertFalse(p("[*.]example.com").matches("https://example.com.evil/"))
        assertTrue(p("[*.]example.com").coversHost("Sub.Example.com."))
        assertFalse(p("example.com").coversHost("sub.example.com"))
    }

    @Test
    fun `narrows to a scheme or a port when the pattern names one`() {
        assertTrue(p("https://example.com").matches("https://example.com/"))
        assertFalse(p("https://example.com").matches("http://example.com/"))
        assertTrue(p("https://example.com").matches("wss://example.com/"))
        assertTrue(p("example.com:8443").matches("https://example.com:8443/"))
        assertFalse(p("example.com:8443").matches("https://example.com/"))
        assertTrue(p("example.com").matches("https://example.com:8443/"))
        assertTrue(p("example.com:443").matches("https://example.com/"))
        assertTrue(p("[::1]").matches("http://[::1]:3000/"))
        assertFalse(p("example.com").matches("about:blank"))
    }

    @Test
    fun `orders from the most specific and picks that one of a list`() {
        val order = listOf(
            "https://www.example.com:8443", "https://www.example.com", "www.example.com:8443", "www.example.com",
            "example.com", "https://[*.]www.example.com", "[*.]www.example.com:8443", "[*.]www.example.com",
            "[*.]example.com"
        )
        assertEquals(order, order.reversed().map { p(it) }.sorted().map { it.text })
        assertTrue(p("a.example") < p("b.example"))
        assertEquals(0, p("a.example").compareTo(p("a.example")))

        val list = listOf("[*.]example.com", "www.example.com", "garbage/", "https://[*.]example.com")
        assertEquals("www.example.com", SitePattern.match(list, "https://www.example.com/")?.text)
        assertEquals("https://[*.]example.com", SitePattern.match(list, "https://cdn.example.com/")?.text)
        assertEquals("[*.]example.com", SitePattern.match(list, "http://cdn.example.com/")?.text)
        assertNull(SitePattern.match(list, "https://other.example/"))
        assertNull(SitePattern.match(list, "about:blank"))
        assertNull(SitePattern.match(emptyList(), "https://example.com/"))
    }
}
